File: rate_limiter/leaky_bucket/leaky_bucket.py

## We need a bucket, so a data structure to store our requests.

## we need a rate

##we need a function that manages the bucket
import time


bucket = []

rate = 2 ## tow by second

capacity = 5 ## max requests



def newRequests(request):
    if len(bucket) > 5:
        print('ignore request:', request)
    else:
        print('accept request:', request)
        bucket.append(request)
        

def processRequest():
    if len(bucket) > 0:
       del bucket[:2]
    else:
        print('empty bucket')


for i in range(7):
    newRequests(i)

while len(bucket)>1:
    time.sleep(1)
    print(bucket)
    processRequest()


            

File: rate_limiter/leaky_bucket/leaky_bucket.js

let bucket = [];
let rate = 2;
let capacity = 5;

function processRequests() {
    if (bucket.length > 0) {
        bucket.splice(0, rate);
    }
}

function newRequests(request) {
    if (bucket.length > capacity) {
        console.log('ignore request:', request);
        return;
    }
    bucket.push(request)
}


for (let x = 0; x < 7; x++) {
    newRequests(x);

}


let timer = setInterval(() => {
    console.log('processing requests', bucket);
    processRequests();
    if (bucket.length === 0) {
        clearInterval(timer);
    }
}, 1000);

for (let x = 0; x < 7; x++) {
    newRequests(x + 2);

}


